sympy/core/basic.py

```diff
@@ -1228,31 +1228,90 @@ def has(self, *patterns):
         False
 
         """
-        return any(self._has(pattern) for pattern in patterns)
+        return self._has(iterargs, *patterns)
 
-    def _has(self, pattern):
-        """Helper for .has()"""
-        from .function import UndefinedFunction, Function
-        if isinstance(pattern, UndefinedFunction):
-            return any(pattern in (f, f.func)
-                       for f in self.atoms(Function, UndefinedFunction))
+    @cacheit
+    def has_free(self, *patterns):
+        """return True if self has object(s) ``x`` as a free expression
+        else False.
 
-        if isinstance(pattern, BasicMeta):
-            subtrees = _preorder_traversal(self)
-            return any(isinstance(arg, pattern) for arg in subtrees)
+        Examples
+        ========
 
-        pattern = _sympify(pattern)
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> g = Function('g')
+        >>> expr = Integral(f(x), (f(x), 1, g(y)))
+        >>> expr.free_symbols
+        {y}
+        >>> expr.has_free(g(y))
+        True
+        >>> expr.has_free(*(x, f(x)))
+        False
 
-        _has_matcher = getattr(pattern, '_has_matcher', None)
-        if _has_matcher is not None:
-            match = _has_matcher()
-            return any(match(arg) for arg in _preorder_traversal(self))
-        else:
-            return any(arg == pattern for arg in _preorder_traversal(self))
+        This works for subexpressions and types, too:
+
+        >>> expr.has_free(g)
+        True
+        >>> (x + y + 1).has_free(y + 1)
+        True
+
+        """
+        return self._has(iterfreeargs, *patterns)
+
+    def _has(self, iterargs, *patterns):
+        # separate out types and unhashable objects
+        type_set = set()  # only types
+        p_set = set()  # all patterns so Tuple(f, Add).has(f) will pass
```

Comment:
The comment suggests that `Tuple(f, Add).has(f)` is a supported use case but it shouldn't be.

Reply:
The `_has` function still seems more complicated than I think it should be. How is this better/different than the version that I proposed?

Reply:
> Tuple(f, Add)...shouldn't be [supported]

It would be backward compatible to not support it. If we are going to drop support I think there should be a broader discussion. In python you can store anything in a list; why can't we store any SymPy object in a Tuple? I don't know what a user wants to do with such a Tuple nor what type of object they might want to make that uses it, but I don't see a reason to limit their options by removing this possibility. Maybe they want to make an object that generates random equations so they want to store different operators in the Tuple.

> How is it different [that the one at https://github.com/sympy/sympy/pull/22539#discussion_r758286349]

1. it supports currently supported syntax (`Tuple(f, Add)`)
2. it supports the dual-nature of `E**x` vs `exp(x)` for which `isinstance(E**x, exp)` is True
2. it checks for exact matches before using `has_match` even for things that have a matcher
3. it recognizes objects that are not hashable
5. it ignores non-Basic types (yours raises an error which is probably better)
6. it sympifies only things that are not Basic (so avoids that path with an `isinstance` test)
7. it uses `any(i in set_types for i in iterargs(self))` instead of any(instance(i, types) for i in set(iterargs(self)))`; I'm not sure of the performance difference there, however. I know that `i in tuple` is slower than `i in set`.

Reply:
> In python you can store anything in a list; why can't we store any SymPy object in a Tuple?

The question is what is a "SymPy object"? The args of Basic should be Basic (instances not subclasses). That is an assumption that is implicit in almost all Basic routines e.g.:
```
In [86]: Tuple(f, Add).subs(x, y)
---------------------------------------------------------------------------
...
~/current/sympy/sympy.git/sympy/core/basic.py in fallback(self, old, new)
   1053                 if not hasattr(arg, '_eval_subs'):
   1054                     continue
-> 1055                 arg = arg._subs(old, new, **hints)
   1056                 if not _aresame(arg, args[i]):
   1057                     hit = True
...
TypeError: _subs() missing 1 required positional argument: 'new' 
```
There are a few Basic routines that have workarounds to try and support broken subclasses that have non-Basic args but those broken classes will always be unreliable and buggy. Instead of trying to work around these things we should do the opposite: ensure that use of non-Basic args blows up as soon as possible. We should also remove anything from the test suite that gives the impression that such cases should be supported so that contributors are not tempted to add workarounds.

@ThePauliPrinciple has been working (in a range of different PRs) on trying to eliminate all cases of non-Basic args. e.g. #22551

Of course it would be nice if an undefined function was a Basic instance (#4787).

Reply:
>  why can't we store any SymPy object in a Tuple

Is this because of the invariant that we are aiming for: `obj.func(*obj.args)`. I know @ThePauliPrinciple has been working on this (as I recall); but special handling is needed for Atom anyway and I'm not sure that special handling shouldn't be done for classes, too, in a detached state.

Reply:
There are two separate invariants:

1. `obj.func(obj.args)` should recreate `obj` (or evaluate to something equivalent to `obj`)
2. All args are Basic.

It is necessary to assume that all args are Basic because Basic is a very complicated class and has two many methods and attributes for anything to meaningfully duck-type as Basic:
```python
In [2]: len(Basic.__dict__)
Out[2]: 127
```
Wherever we have non-Basic in the args we have bugs. Wherever we have tests that assert that we can have non-Basic in the args it forces contributors into workarounds that degrade the quality, reliability and performance of routines that would be quite simple if the two invariants above could be relied upon.